### PR TITLE
A simple GM Halo2 circuit for testing

### PR DESCRIPTION
**Pull request by 0xisk**
### Changes in this PR: 
1. Fork a version from the origin impl of GM encryption (https://github.com/crodriguezvega/probabilisticpubkey) and update it's dependence's versions.
2. Impl a simple halo2 circuit using `halo2-lib` and `halo2-wasm`.

### CODE REVIEW

File: .vscode/settings.json
Comment (ndrwnaguib) [inline .vscode/settings.json:1]: Shouldn't this be in `.gitignore`?

File: packages/circuits/configs/gm_encryption.config
Comment (ndrwnaguib) [inline packages/circuits/configs/gm_encryption.config:1]: A few comments on how these constants were chosen would be useful.

File: packages/circuits/src/gm_encryption.rs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    //     halo2_wasm.config(circuit_params);
+
+    //     let mut circuit = GmVerificationCircuit::new(&halo2_wasm, gm_verification_inputs);
+
+    //     circuit.verify_encryption();
+
+    //     halo2_wasm.mock();
+
+    //     Ok(())
+    // }
```
Comment (ndrwnaguib) [inline packages/circuits/src/gm_encryption.rs:172]: This snippet seems to be the same as the one above it? Was it left by mistake or we are keeping it for a reason?

**Review by ndrwnaguib — CHANGES_REQUESTED** [review]
Thank you @0xisk. The PR looks good to me, I left a few comments. Also, would it be better if we add `probabilistic-encryption` as a submodule instead since we are not modifying the code in anyway?

File: packages/circuits/configs/gm_encryption.config
Comment (0xisk) [reply packages/circuits/configs/gm_encryption.config:1]: These are the default values commonly used in various Halo2 examples. However, I plan to add a detailed explanation of how these values are chosen next week. I'm currently benchmarking another Halo2 project, which will provide deeper insights into the rationale behind these selections.

File: packages/circuits/src/gm_encryption.rs (hunk 1 above)
Comment (0xisk) [reply packages/circuits/src/gm_encryption.rs:172]: No it wasn't left by mistake, I was thinking how to get this test successful at the end as the expected constraint is that the `expected_cipher` is not equal to the `actual_cipher`. But I didn't find a way for doing such a constraint yet. So I will remove this test unit for now.

**Comment by 0xisk** [comment]
> @0xisk As we have just decided during the meeting, would it be possible to export the changes as a patch file and add `probabilistic-encryption` as a submodule to which we apply the patch file after recursively cloning the this repo.

@ndrwnaguib I have pushed the changes into a patch file, also have updated the README docs for dealing with submodules.

**Comment by 0xisk** [comment]
After team discussions, we've decided to switch from Halo2 to Risc0. That is because at the moment it seems to us the most ready one and it provides good proving experience (runnable on laptops), and verifying experience (blockchain integration use groth16, same as circom). (@ndrwnaguib)

**Pull request closed without merging**